src/Microsoft.VisualStudio.Services.Agent/Util/VarUtil.cs

```diff
@@ -169,17 +188,44 @@ public static void ExpandValues(IHostContext context, IDictionary<string, string
                         startIndex: prefixIndex + Constants.Variables.MacroPrefix.Length,
                         length: suffixIndex - prefixIndex - Constants.Variables.MacroPrefix.Length);
                     trace.Verbose($"Found macro candidate: '{variableKey}'");
-                    string variableValue;
-                    if (!string.IsNullOrEmpty(variableKey) &&
-                        TryGetValue(trace, source, variableKey, out variableValue))
+
+                    var isVariableKeyPresent = !string.IsNullOrEmpty(variableKey);
+
+                    if (isVariableKeyPresent &&
+                        shellName != WellKnownScriptShell.Unknown &&
+                        shellName != WellKnownScriptShell.Cmd &&
+                        Constants.ScriptShells.EnvVariablePartsPerShell.TryGetValue(shellName, out var shellEnvVariableParts) &&
+                        Constants.Variables.VariablesVulnerableToExecution.Contains(variableKey, StringComparer.OrdinalIgnoreCase)
+                    )
+                    {
+                        var envVariableName = ConvertToEnvVariableFormat(variableKey);
+                        var shellEnvVariable = shellEnvVariableParts.Prefix + envVariableName + shellEnvVariableParts.Suffix;
+
+                        var warningMessage = StringUtil.Loc("VariableVulnerableToExecWarn", variableKey, shellEnvVariable);
+                        warnings.Add(warningMessage);
+                    }
+                    if (isVariableKeyPresent &&
+                        TryGetValue(trace, source, variableKey, out string variableValue))
                     {
                         // A matching variable was found.
                         // Update the target value.
                         trace.Verbose("Macro found.");
+
+                        if (shellName == WellKnownScriptShell.Cmd &&
+                            Constants.Variables.VariablesVulnerableToExecution.Contains(variableKey, StringComparer.OrdinalIgnoreCase)
+                        )
+                        {
+                            trace.Verbose("CMD shell found. Custom macro processing.");
+
+                            // When matching "&", "|", "<" and ">" cmd commands adds "^" before them.
+                            var cmdCommandsRegex = new Regex(@"[&|\||<|>]");
+                            variableValue = cmdCommandsRegex.Replace(variableValue, "^$&");
+                        }
+
                         targetValue = string.Concat(
-                            targetValue.Substring(0, prefixIndex),
-                            variableValue ?? string.Empty,
-                            targetValue.Substring(suffixIndex + Constants.Variables.MacroSuffix.Length));
+                            targetValue[..prefixIndex],
```

Comment:
>targetValue [](http://example.com/codeflow?start=28&length=11)

[suggestion/not blocking]

It looks like this is just a style change, not functional?

It's easier to review if you keep consistent style, especailly for index access as we'd want to review off-by-one errorrs closely

Substring is more often used when manipulating strings than ranges

if there is no functional change my suggestion is to leave it as-is

Reply:
> Should I also revert the minor formatting changes?

I personally avoid even minor formatting changes on unchanged code, (even white space to avoid extra diffs) to make review easier.  Changing operators adds additional distraction (and risk for potential mistakes), even if the refactoring was done safely using tooling, the reviewer can't tell that's the case

Reply:
@merlynomsft I have prepared another PR that should remove the formatting differences in this one, please check - https://github.com/microsoft/azure-pipelines-agent/pull/4101

Reply:
Thanks for opening a PR; I like the approach!

If you haven't already, please carefully review the tooling output to sanity in case of tooling bugs, especially since it's a large bulk change; also consider it may cause conflicts in this PR and others that need to be resolved. 

I think we can close this comment; no need to reopen if you take a different approach.